src/librustc_mir/borrow_check/borrow_set.rs

```diff
@@ -116,19 +113,6 @@ impl<'tcx> BorrowSet<'tcx> {
             visitor.visit_basic_block_data(block, block_data);
         }
 
-        // Double check: We should have found an activation for every pending
-        // activation.
-        assert_eq!(
-            visitor
-                .pending_activations
-                .iter()
-                .find(|&(_local, &borrow_index)| visitor.idx_vec[borrow_index]
-                    .activation_location
-                    .is_none()),
-            None,
-            "never found an activation for this borrow!",
-        );
-
```

Comment:
I am a bit concerned here because I don't know where two-phase borrows that wind up unactivated will be set to `TwoPhaseActivation::NotActivated`. I think they will erroneously be recorded as not two phase. This means we will get an error with e.g. this code:

```rust
fn main() {
    let mut vec = vec![];
    vec.push((vec.len(), panic!()));
}
```

I expect this to error because the borrow of `vec` for calling `push` will incorrectly be considered "not two phase".

Reply:
To fix this, I think we can modify `insert_as_pending_if_two_phase` to set the `activation_location` field to `NotActivated` instead of `NotTwoPhase`. Then when we *do* find an activation we can assert that the field is "not activated".

Reply:
If I am correct about all this btw please add that above test. =)

Reply:
Pushed a commit that addresses this.